Kernel/Devices/Storage/USB/BulkSCSIInterface.cpp

```diff
@@ -16,6 +16,109 @@ BulkSCSIInterface::BulkSCSIInterface(LUNAddress logical_unit_number_address, u32
     , m_in_pipe(move(in_pipe))
     , m_out_pipe(move(out_pipe))
 {
+    // Note: If this fails, it only means that we may be inefficient in our way
+    //       of talking to the device.
+    (void)query_characteristics();
+}
+
+ErrorOr<void> BulkSCSIInterface::query_characteristics()
+{
+    SCSI::Inquiry inquiry_command {};
+    inquiry_command.enable_vital_product_data = 1;
+    u8 vital_product_page_buffer[0xfc];
+    SCSI::SupportedVitalProductPages& vital_product_page = *reinterpret_cast<SCSI::SupportedVitalProductPages*>(vital_product_page_buffer);
+
+    inquiry_command.page_code = SCSI::VitalProductDataPageCode::SupportedVitalProductDataPages;
+    inquiry_command.allocation_length = sizeof(vital_product_page_buffer);
+
+    auto status = TRY(send_scsi_command<SCSIDataDirection::DataToInitiator>(*m_out_pipe, *m_in_pipe, inquiry_command, &vital_product_page, sizeof(vital_product_page_buffer)));
+
+    if (status.status != CSWStatus::Passed) {
+        dbgln("SCSI/BBB: Inquiry failed to inquire supported vital product data pages with code {}", to_underlying(status.status));
+        // FIXME: Maybe request sense here
+        // FIXME: Treating this as an error for now
+        // Some HW seems to stall this and/or send garbage...
+        return EIO;
+    }
+    if (vital_product_page.page_code != SCSI::VitalProductDataPageCode::SupportedVitalProductDataPages) {
+        dmesgln("SCSI/BBB: Returned wrong page code for supported vital product data pages: {:#02x}", to_underlying(vital_product_page.page_code));
+        return EIO;
+    }
+
+    if ((vital_product_page.page_length + 3uz) > sizeof(vital_product_page_buffer)) {
```

Comment:
Is the PAGE LENGTH inclusive now or not? If it is, shouldn't this be 4uz?

Reply:
Yeah lets make it `4uz`, although not sure if we ever gonna hit that

Reply:
For the BlockLimits Page `sizeof(...)-4` seems to be correct

Reply:
And for the supported VPD page as well?

Reply:
For that we the 4 should be fine as that would only mean that we fire off the warning an entry early, and as the comment notes, this should be impossible to reach anyway, afaict

Reply:
Right, sorry I just got confused there for a moment.